src/v/kafka/protocol/schemata/generator.py

```diff
@@ -1128,6 +1148,51 @@ def is_error_code(self):
 {%- endif %}
 {% endmacro %}
 
+{% macro render_formatter(struct) %}
+{%- if not struct.is_streamable %}
+template<>
+struct fmt::formatter<kafka::{{ struct.name }}, char> {
+
+    template <typename ParseContext>
+    constexpr auto parse(ParseContext &ctx) -> decltype(ctx.begin()) {
+        return ctx.begin();
+    }
+
+    template <typename FormatContext>
+    auto format([[maybe_unused]] const kafka::{{ struct.name }} &v, FormatContext& ctx) const
+            -> decltype(ctx.out()) {
+        return fmt::format_to(ctx.out(),
+          "{{'{{' + struct.format + '}}'}}",
+          {%- for field in struct.fields %}
+          {%- if field.is_sensitive %}"****"{% else %}v.{{ field.name }}{% endif %}{% if not loop.last %},{% endif %}
+
+          {%- endfor %}
+        );
+    }
+};
+{%- for field in struct.fields %}
+{%- if field.is_array and not field.is_streamable %}
+{%- set vtype = field.value_type %}
+{%- set container = (field.type_name_parts() | list)[1] %}
+template<>
+struct fmt::formatter<{{ container }}<kafka::{{ vtype }}>> {
+    template<typename ParseContext>
+    constexpr auto parse(ParseContext& ctx) -> decltype(ctx.begin()) {
+        return ctx.begin();
+    }
+
+    template<typename FormatContext>
+    auto format(
+      [[maybe_unused]] const {{ container }}<kafka::{{ vtype}}>& v,
+      FormatContext& ctx) const -> decltype(ctx.out()) {
+        return fmt::format_to(ctx.out(), "{}", fmt::join(v, ","));
+    }
+};
+{%- endif %}
+{%- endfor %}
```

Comment:
This can be replaced by
```c++
#include <fmt/ranges.h>
```

Reply:
🤯

Reply:
This seems to introduce some ambiguity in overload resolution. Maybe due to including both `ostream.h` and `ranges.h` in the same compilation unit?

Reply:
That's strange. I'm sure I compiled it.

Reply:
Sorry, forgot to specify "on the OSS build", which uses fmtlib 8.1.1. Did not get around to trying a more recent fmtlib version, but could do.

Reply:
Oh, this failure?
```
/usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/format:3824:5: note: candidate function [with _Out = fmt::appender, _Args = <unsigned int &>]
    format_to(_Out __out, format_string<_Args...> __fmt, _Args&&... __args)
    ^
/__w/redpanda/redpanda/build/release-ci/_deps/fmt-src/include/fmt/core.h:3146:17: note: candidate function [with OutputIt = fmt::appender, T = <unsigned int &>, $2 = 0]
FMT_INLINE auto format_to(OutputIt out, format_string<T...> fmt, T&&... args)
```
So somehow the C++ 20 format is being pulled into scope from libstdc++ 13

Reply:
oh...yeah I didn't parse out the path there. that's weird

Reply:
There are (or have been historically) several cases of this in `fmtlib`. Cost of standardization I guess. The problem is that calls to `format_to` in `ranges.h` are not namespace qualified, so we get ADL'ed into ambiguity with `std::format_to`. 

One such discussion: https://github.com/fmtlib/fmt/issues/2920. Unrelated to our specific case, which I suspect is resolved painlessly in future versions (the offending calls are no longer present in fmtlib/fmt:master).

Bad lookups for reference (`fmt/ranges.h:L488-509`)

```c++
if (is_utf8()) {
  if (escape.cp < 0x100) {
    out = format_to(out, "\\x{:02x}", escape.cp);
    continue;
  }
  if (escape.cp < 0x10000) {
    out = format_to(out, "\\u{:04x}", escape.cp);
    continue;
  }
  if (escape.cp < 0x110000) {
    out = format_to(out, "\\U{:08x}", escape.cp);
    continue;
  }
}
for (Char escape_char : basic_string_view<Char>(
         escape.begin, to_unsigned(escape.end - escape.begin))) {
  out = format_to(
      out, "\\x{:02x}",
      static_cast<typename std::make_unsigned<Char>::type>(escape_char));
}
```

Reply:
Not quite sure how to proceed with this one. It looks like several other compilation units include `fmt/ranges.h`, but maybe we're struggling here due to the catch-all nature of the generated source file.

Tempting to let sleeping dogs lie, though, as you suggested earlier, we could reasonably try upgrading fmtlib. Not sure how much work that would be.

Reply:
Upgrading libfmt is a big job. Happy to leave this as-is.

Reply:
coo. patch with `format_string` checks incoming